src/main/java/hudson/plugins/gradle/injection/BuildScanEnvironmentContributor.java

```diff
@@ -9,29 +10,78 @@
 import hudson.model.Run;
 import hudson.model.TaskListener;
 import hudson.plugins.gradle.DevelocityLogger;
+import hudson.plugins.gradle.injection.token.ShortLivedTokenClient;
 import hudson.util.Secret;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.net.MalformedURLException;
+import java.net.URL;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 @Extension
 public class BuildScanEnvironmentContributor extends EnvironmentContributor {
 
+    private final ShortLivedTokenClient tokenClient;
+
+    public BuildScanEnvironmentContributor() {
+        this.tokenClient = new ShortLivedTokenClient();
+    }
+
+    public BuildScanEnvironmentContributor(ShortLivedTokenClient tokenClient) {
+        this.tokenClient = tokenClient;
+    }
+
     @Override
     public void buildEnvironmentFor(@Nonnull Run run, @Nonnull EnvVars envs, @Nonnull TaskListener listener) {
         Secret secretKey = InjectionConfig.get().getAccessKey();
         Secret secretPassword = InjectionConfig.get().getGradlePluginRepositoryPassword();
         if ((secretKey == null && secretPassword == null) || alreadyExecuted(run)) {
             return;
         }
+        DevelocityLogger logger = new DevelocityLogger(listener);
+
+        Secret shortLivedToken = getShortLivedToken(secretKey, logger);
+
+        run.addAction(DevelocityParametersAction.of(logger, shortLivedToken, secretPassword));
+    }
+
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
+    private @Nullable Secret getShortLivedToken(Secret secretKey, DevelocityLogger logger) {
+        if (secretKey == null) {
+            return null;
+        }
+        String serverUrl = InjectionConfig.get().getServer();
+        String hostname = getHostnameFromServerUrl(serverUrl);
+        if (hostname == null) {
+            logger.error("Could not extract hostname from Develocity server URL");
+            return null;
+        }
+        if (!DevelocityAccessKey.isValid(secretKey.getPlainText())) {
+            logger.error("Develocity access key format is not valid");
+            return null;
+        }
+        Optional<DevelocityAccessKey> accessKey = DevelocityAccessKey.parse(secretKey.getPlainText(), hostname);
+        return accessKey
+            .map(k ->
+                tokenClient.get(serverUrl, k, InjectionConfig.get().getShortLivedTokenExpiry()))
+            .filter(Optional::isPresent)
+            .map(k -> Secret.fromString(k.get().getRawAccessKey()))
+            .orElse(null);
```

Comment:
If we fail to get a short-lived access token, do we want to fail the build, or fall back to some reasonable behavior?

In our own dogfooding case, we decided to not set the access key and let the build fail to use various Develocity features like build caching and build scan publishing. However, we had to explicitly disable Test Distribution (which will otherwise fail if it fails to connect to Develocity). I wonder if we should do something similar here.

I also wonder if it would make sense to add a configuration option to allow users to decide if they want the build to fail if it can't get an access token, or to continue without Develocity.